lib/msprime.c

```diff
@@ -4950,73 +5055,57 @@ msp_set_simulation_model_beta(msp_t *self, double population_size, double alpha,
 {
     int ret = 0;
-    size_t j;
     uint32_t locus;
     simulation_model_t *model = &self->model;
+    genic_selection_trajectory_t *trajectory =
+        &model->params.sweep.trajectory_params.genic_selection_trajectory;
     double L = self->recomb_map->sequence_length;
 
     /* Check the inputs to make sure they make sense */
     if (position < 0 || position >= L) {
         ret = MSP_ERR_BAD_SWEEP_POSITION;
         goto out;
     }
-    if (num_steps == 0) {
-        ret = MSP_ERR_EMPTY_TRAJECTORY;
+    if (start_frequency < 0.0 || start_frequency > 1.0
+            || end_frequency < 0.0 || end_frequency > 1.0) {
+        ret = MSP_ERR_BAD_ALLELE_FREQUENCY;
         goto out;
     }
-    for (j = 0; j < num_steps; j++) {
-        if (j > 0) {
-            if (time[j - 1] >= time[j]) {
-                ret = MSP_ERR_BAD_TRAJECTORY_TIME;
-                goto out;
-            }
-        }
-        if (time[j] < 0) {
-            ret = MSP_ERR_BAD_TRAJECTORY_TIME;
-            goto out;
-        }
-        if (allele_frequency[j] < 0 || allele_frequency[j] > 1) {
-            ret = MSP_ERR_BAD_TRAJECTORY_ALLELE_FREQUENCY;
-            goto out;
-        }
+    if (start_frequency >= end_frequency) {
+        ret = MSP_ERR_BAD_TRAJECTORY_START_END;
+        goto out;
+    }
+    if (dt <= 0) {
+        ret = MSP_ERR_BAD_TIME_DELTA;
+        goto out;
     }
+    /* TODO any limitations on alpha? */
 
     ret = recomb_map_phys_to_discrete_genetic(self->recomb_map, position, &locus);
     if (ret != 0) {
         goto out;
     }
-    ret = msp_set_simulation_model(self, MSP_MODEL_SINGLE_SWEEP, population_size);
+    ret = msp_set_simulation_model(self, MSP_MODEL_SWEEP, population_size);
     if (ret != 0) {
         goto out;
     }
-    self->model.free = single_sweep_model_free;
-    model->params.single_sweep.locus = locus;
-    model->params.single_sweep.trajectory.num_steps = num_steps;
-    model->params.single_sweep.trajectory.allele_frequency = malloc(
-            num_steps * sizeof(double));
-    model->params.single_sweep.trajectory.time = malloc(num_steps * sizeof(double));
-    if (model->params.single_sweep.trajectory.allele_frequency == NULL
-            || model->params.single_sweep.trajectory.time == NULL) {
-        ret = MSP_ERR_NO_MEMORY;
-        goto out;
-    }
-    memcpy(model->params.single_sweep.trajectory.allele_frequency,
-            allele_frequency, num_steps * sizeof(double));
-    memcpy(model->params.single_sweep.trajectory.time,
-            time, num_steps * sizeof(double));
-    /* TODO we need to rescale the times into model units here */
+    model->params.sweep.locus = locus;
+    model->params.sweep.generate_trajectory = genic_selection_generate_trajectory;
+    model->params.sweep.print_state = genic_selection_print_state;
+    trajectory->start_frequency = start_frequency;
+    trajectory->end_frequency = end_frequency;
+    /* JK: question for Andy: what are the dimensions of alpha? Does it
```

Comment:
\alpha = 2Ns. no reason to rescale it.

Reply:
Ah, OK. For consistency I think we should change the top-level parameter to s. There are no other scaled parameters in the Python API, and s is a lot more user-friendly than alpha. I'll leave it for the moment though. and merge this now so you can get working.